Delegated Opt In Review (#1)
joe-p/delegated_optin_arc: README.md

## Motivation
[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control whether the lsig is still functionality or not. Without an application to control this, signing the logic signature program would be irreversible without rekeying. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt-ins.

## Specification
This is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). There is additional functionality provided by the application in this ARC in the methods of the application.

For all methods, refer to the [ABI JSON description](./contracts/artifacts/DelegatedOptIn.abi.json) for complete descriptions of arguments.

### Signature Storage

> `setSignature(byte[64],address,txn)void` is a method that allows a user to upload their signature to box storage

Reviewer: Is it really `setSignature(byte[64],address,pay)void`

(txn -> pay)

Author: Updated in b69d790b048c5682fd8549e063b609651f01bb4c

### Opt-Ins

`delegatedOptIn(pay,axfer)void` is an implementation of the [ARCX](https://github.com/algorandfoundation/ARCs/pull/229) interfaces. It verifies the MBR payment is sent to the account opting in and that it covers the ASA minimum balance requirement.

### Storage

| Type | Key | Description |
| ---- | --- | ----------- |
| Box | `auth-addr` | Mapping of signer to lsig signature |

## Rationale
Box storage is used to store signatures indefinitely and [the verifier lsig](./contracts/verifier_lsig.teal) ensures the signature is always correct.

> End time functionality is provided to allow users to revert the effects of signing the logic signature programs without having to rekey. It also lets users open their account for opt-in delegations for a short amount of time without having to remember to manually undo it.

Reviewer 2: This should be updated

Author: Updated in b69d790b048c5682fd8549e063b609651f01bb4c

Asset MBR is stored in global storage in case the MBR for asset were to ever change.

## Backwards Compatibility
N/A

## Test Cases
Tests written with Jest and algokit for the contract and logic signatures can be seen in [test.ts](./tests/test.ts).

```
  Delegated Opt In App
    create
      ✓ creates the app (765 ms)
    setSignature
      ✓ works with valid signature and lsig (790 ms)
    delegatedOptIn
      ✓ works with valid lsig and method call (1453 ms)

>     unsetSignature

Reviewer: How about "revokeSignature" instead?

Author: Yeah I like that better. Changed in 89e1515a7c69e88f5370735d8d516379549fbc77

      ✓ sends back MBR (753 ms)
      ✓ doesn't allow opt-ins (1186 ms)
```

## Reference Implementation
[delegated_optin_app.algo.ts](./contracts/delegated_optin_app.algo.ts) is the application written in TEALScript.

[DelegatedOptIn.approval.teal](./contracts/artifacts/DelegatedOptIn.approval.teal) is the TEAL compiled from TEALScript.

## Security Considerations

The test cases test proper functionality of all methods, but there has been no extended effort in attempt to break the contract. Most of the functionality in the app and logic signatures is relatively simple, so the chances of unexpected bugs is relatively low.

> It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted-in to assets via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the signature will work regardless of the latest timestamp.

Reviewer 2: same here

Author: Updated in b69d790b048c5682fd8549e063b609651f01bb4c

## Abstract

> This ARC contains is implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will deployed on Algorand networks. 

Reviewer: ```suggestion
This ARC contains is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will be deployed on Algorand networks. 
```

> This ARC contains is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will be deployed on Algorand networks. 

Reviewer: I can see why this is a nice, useful app to use in combination with the ARC-50, but I don't see why it should be considered "special". The point of ARC-50 was to give users control - let them have an app that they trust to check opt-ins for them.  This is one of them, but if it's the only one, then the flexibility that ARC-50 promises is reduced to the flexibility that this particular implementation provides.

Author: Yeah the idea isn't that this is the only ARC50 implementation, but it's a standardized one that every wallet/dApp in the ecosystem can integrate. Other people can make whatever implementations they want, but the more implementations there are the more fragmented the ecosystem can get.

## Motivation

> [ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control how long the signature is valid for. Without an application to control this, signing the logic signature program would be irriversible. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt ins.

Reviewer: ```suggestion
[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control how long the signature is valid for. Without an application to control this, signing the logic signature program would be irreversible. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt-ins.
```

## Specification

> This is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). There is additional functionality provided by the applicaiton in this ARC in the methods of the application.

Reviewer: ```suggestion
This is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). There is additional functionality provided by the application in this ARC in the methods of the application.
```

For all methods, refer to the [ABI JSON description](./contracts/artifacts/DelegatedOptIn.abi.json) for complete descriptions of arguments.

### Signature Storage

> `setOpenOptInSignature(byte[64],address,txn)void` and `setAddressOptInSignature(byte[64],address,address,txn)void` are methods for adding signatures to box storage for open opt ins and address opt ins, respectively.

Reviewer: ```suggestion
`setOpenOptInSignature(byte[64],address,txn)void` and `setAddressOptInSignature(byte[64],address,address,txn)void` are methods for adding signatures to box storage for open opt-ins and address opt-ins, respectively.
```

Both methods utilize [the verifier lsig](./contracts/verifier_lsig.teal) to verify the address being added to box storage matches the address(es) used as the key. This means anyone can query the applications boxes to get the signature for the given address(es) and be certain that it is correct.

### End Times

> `setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt ins and ddress opt ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any value.

Reviewer: ```suggestion
`setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt-ins and address opt-ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any value.
```

> `setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt-ins and address opt-ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any value.

Reviewer: Does setting the endtime to 0 delete the box?  Does the MBR get recovered somehow?

Actually, I'm assuming that the non-existence of an endtime means there is no endtime? So perhaps setting the endtime to MaxUint64 should delete the box and recover the MBR.

I suppose that really I'm asking for a way to delete all of the storage and have it mean "I want out of this whole business, give me my MBR back".

Author: Right now there is no MBR recovery mechanism. I figured it added extra complexity for such a small amount of ALGO it wasn't necessarily worth it. Open to the idea of MBR recovery though.

Reviewer: I'm doubling down on my dislike of storing logicsigs on chain without a deletion mechanism.  If I say I want to allow someone to opt me in, I should just be able to remove that.

I suppose I'm really confused by the specific sender mechanism. What's the use case that makes it worth enshrining this idea?  Why not just "allow all, deny all"? (and by that I mean let's also scrap endtime.  just allow revocation.)

Reviewer: Ok, now I see. You are just storing the signature.  That seems very reasonable.

But it also means that using one box for the signature and the endtime is easy. They are both fixed size.

Author: Assuming we still want to maintain the current design goals, we actually couldn't make it one box. The signatures are stored per the resepective public key, but the end times are stored per address. This means one can have multiple accounts rekeyed to a single auth account and have different rules set for each one. 

For example, I could have a single hardware wallet that is the auth addr for a "DeFi Account" and an "NFT Account". The "DeFi Account" delegates opt-ins to DeFi apps and "NFT Account" delegates opt-ins to NFT marketplaces, but not the other way around. This can currently be done because I can set the end times to be unique for each account.

Author: I should mention the downside of this approach is it makes fresh rekeyed accounts open to any opt-ins before setting the end times, but in general I think this is reasonable tradeoff for having per-account control. Open to other opinions though.

Reviewer: > The signatures are stored per the resepective public key, but the end times are stored per address. This means one can have multiple accounts rekeyed to a single auth account and have different rules set for each one.
> 
> For example, I could have a single hardware wallet that is the auth addr for a "DeFi Account" and an "NFT Account". The "DeFi Account" delegates opt-ins to DeFi apps and "NFT Account" delegates opt-ins to NFT marketplaces, but not the other way around. This can currently be done because I can set the end times to be unique for each account.

I think it's probably a mistake to have this separation. Today, rekeying two accounts to the same key means that they are under the exact same authority.  It seems like a bad idea to introduce a mechanism that changes that model. If this becomes popular, we then have two ways of thinking about rekeying, and users must understand both.

Author: Yeah true that's a really good point. Removing this means we can just use the existence of the signature boxes to signify whether delegated opt ins are permitted or not. We can also remove the time component and if people really want an automated disable, that can be done at a higher level.

> ### Opt Ins

Reviewer: ```suggestion
### Opt-Ins
```

> `openOptIn(pay,axfer)void` and `addressOptIn(pay,axfer)void` are implementations of the [ARCX](https://github.com/algorandfoundation/ARCs/pull/229) interfaces. They both verify the MBR payment is sent to the account opting in and that it covers the ASA minimum balance requirement, which is stored in global storage. It also verifies the value of `global LatestTimestamp` is less than the set end time for the account opting in.

Reviewer: By aware, though I don't think it matters, that `global LatestTimestamp` is for the _last_ block.

So if the last block was at 9998, and the end time is 10000, but the next block gets added at 10002, then you can have an opt-in the occurs after the set endtime.

Author: Yeah that's why I specified the usage of `global LatestTimestamp`. The two options are either `Round` or `LatestTimestamp`. I feel like the timestamp approach allows for a better UX even though there may be some slight deviations in the true end time/block.

### Storage

| Type | Key | Description |
| ---- | --- | ----------- |
| Global | "sigVerificationAddress" | Stores the address of [the verifier lsig](./contracts/verifier_lsig.teal) |
| Global | "assetMBR" | Stores the ASA MBR |
| Box | "s-" + `auth-addr` | Mapping of signer to open opt-in signature |
| Box |  "e-" + `auth-addr` | Mapping of signer to open opt-in end time |

> | Box | "s-" + `sha256(signer,sender)` | Mapping of the hash of the signer address plus sender address to address opt-in signature |

Reviewer: Using the hash, rather than concatenation saves little space (and MBR), but I'd like to understand the intended use, because it also makes some things harder. If an intended use is "allow an exchange to opt me in as needed", using the hash makes it impossible for the exchange to troll through the boxes to learn who it can opt-in for.  (It can check specific addresses, but not find out all the addresses is can opt-in.). To help with that, you might emit an event when such a box is set, which, one presumes, will be indexed by interested parties.

Author: The hash is used because there are two key-value pairs that use the `(sender,signer)` tuple as a key (end times and signature), thus a prefix is needed to differentiate them which results in more than 64 bytes. 

There could be a single box that stores both the signature and the end time under a single key, but that does add some extra complexity to decoding that data structure (vs just checking box existence).

Reviewer: If the signature box got to be the one named by the concatenation, then you'd enable the kind of lookup I'm describing.

Author: Ah true didn't think of that. Will update so the key for signatures is just the non-hashed `(sender,signer)` tuple

| Box |  "e-" + `sha256(signer,sender)` | Mapping of the hash of the signer address plus sender address to address opt-in end time |

## Rationale
Box storage is used to store signatures indefinitely and [the verifier lsig](./contracts/verifier_lsig.teal) ensures the signature is always correct.

> End time functionality is provided to allow users to revert the effects of signign the logic signature programs without having to rekey. It also let's users open their account for opt-in delegations for a short amount of time without having to remember to manually undo it.

Reviewer: ```suggestion
End time functionality is provided to allow users to revert the effects of signing the logic signature programs without having to rekey. It also lets users open their account for opt-in delegations for a short amount of time without having to remember to manually undo it.
```

      ✓ works with valid address (939 ms)
    setOpenOptInSignature
      ✓ works with valid signature and lsig (1031 ms)
    openOptIn
      ✓ works with valid lsig and method call (973 ms)
    setOpenOptInEndTime
      ✓ works with 0xffffffff as the end time (2091 ms)
      ✓ works with 0 as the end time (1137 ms)
    setAddressOptInSignature
      ✓ works with valid signature and lsig (926 ms)
    addressoptOptIn
      ✓ works with valid lsig, method call, and sender (914 ms)
    setAddressOptInEndtime
      ✓ works with 0xffffffff as the end time (1001 ms)
      ✓ works with 0 as the end time (1611 ms)
```

## Reference Implementation
[delegated_optin_app.algo.ts](./contracts/delegated_optin_app.algo.ts) is the application written in TEALScript.

[DelegatedOptIn.approval.teal](./contracts/artifacts/DelegatedOptIn.approval.teal) is the TEAL compiled from TEALScript.

## Security Considerations

The test cases test proper functionality of all methods, but there has been no extended effort in attempt to break the contract. Most of the functionality in the app and logic signatures is relatively simple, so the chances of unexpected bugs is relatively low.

> It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted in to asset via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the signature will work regardless of the latest timestamp.

Reviewer: ```suggestion
It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted-in to assets via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the signature will work regardless of the latest timestamp.
```

This ARC contains is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will be deployed on Algorand networks. 

## Motivation

> [ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control whether the lsig is still functionality or not. Without an application to control this, signing the logic signature program would be irreversible without rekeying. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt-ins.

Reviewer 2: ```suggestion
[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control whether the lsig is still functional or not. Without an application to control this, signing the logic signature program would be irreversible without rekeying. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt-ins.
```

small typo
